### ISSUE TITLE
Add a way to disable a plugin

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#14026]
**The version of ESLint you are using.**

v7.17.0

**The problem you want to solve.**

Disable a plugin in a shareable config.

**Your take on the correct solution to problem.**

Native support for the equivalent functionality of https://www.npmjs.com/package/eslint-plugin-disable.

e.g.

```json
{
  "plugins": ["react", "disable"],
  "processor": "disable/disable",
  "overrides": [
    {
      "files": ["tests/**/*.test.js"],
      "settings": {
        "disable/plugins": ["react"]
      }
    }
  ]
}

```

**Are you willing to submit a pull request to implement this change?**

No.

CC @mradionov

**Comment by ljharb** [comment]
This seems like it would allow consumers of, say, `eslint-config-airbnb` to disable all accessibility rules en masse, which is explicitly something I don't want them to be able to do. The plugin used is an implementation detail (and i'd argue a proper shared config already specifies them in `plugins`, so that consumers don't have to), and it shouldn't be possible for shared config users to couple tightly to the implementation details of the config.

(As for the performance cost, that's a one-time cost paid for building up the dependency graph, and if you disable all of the rules that use it, the hit will go away. In practice, you can use https://www.npmjs.com/package/jest-runner-eslint to completely remove the cost of this, as airbnb did internally a few years ago)

What you can do is enumerate over the rules in the plugin, and explicitly disable them in a `.js` config, but I really hope that doesn't become easy to do built-in.

**Comment by gajus** [comment]
> This seems like it would allow consumers of, say, eslint-config-airbnb to disable all accessibility rules en masse, which is explicitly something I don't want them to be able to do.

This seems extremely fitted around one person's use case / desires.

**Comment by ljharb** [comment]
That's the whole point - the owner of the shared config's desires are the ones the consumer wants to fit around. If you don't want that, you're free not to use that shared config.

**Comment by gajus** [comment]
> (As for the performance cost, that's a one-time cost paid for building up the dependency graph, and if you disable all of the rules that use it, the hit will go away. In practice, you can use https://www.npmjs.com/package/jest-runner-eslint to completely remove the cost of this, as airbnb did internally a few years ago)

That does not solve problem for IDEs. Just `import` plugin is causing 200ms delay saving a single file with VSCode ESLint plugin enabled.

**Comment by gajus** [comment]
> That's the whole point - the owner of the shared config's desires are the ones the consumer wants to fit around. If you don't want that, you're free not to use that shared config.

I am the owner of `canonical`, and even in my own project, I don't see an option to disable just the plugin that I wish to disable.

I could breakdown `canonical` into many smaller configs and ask users to selectively enable plugins, but that severely impacts usability of the config.

**Comment by ljharb** [comment]
I can't stop people from using eslint-plugin-disable, but that doesn't make the point moot - it'd be a lot easier for people to do it casually if it were built-in, and it shouldn't be easier to do this.

**Comment by gajus** [comment]
If anyone needs this functionality, in a way that disables the plugins, not just the tests (like `eslint-plugin-disable`), then you may want to try https://www.npmjs.com/package/eslint-disable.

```js
const {
  extractConfig,
  disablePlugins,
} = require('eslint-disable');

const baseConfig = {
  'extends': [
    'canonical'
  ],
  'root': true,
};

// This will disable "import" plugin and all rules matching "import/*" pattern.
module.exports = disablePlugins(
  extractConfig(
    baseConfig,
  ),
  [
    'import',
  ]
);

```

I still think this should be in the core of ESLint.

**Comment by nzakas** [comment]
@gajus can you explain succinctly what you mean by "disable a plugin"? That sentence alone is not really enough to evaluate this proposal.

**Comment by ljharb** [comment]
I assume what they want is a way to cancel out the presence of a plugin in a “plugins” list, as well as setting all rules from that plugin to “off”, such that effectively no code from a plugin is evaluated (obv some would be, since plugin code can create configs, and it might be required somewhere - but the rules wouldn’t be invoked)

**Comment by anikethsaha** [comment]
as far as the rules from plugins are concerned, Maybe support for glob patterns in rules can be helpfull (if not already supported) example 

```js
"rules" : {
   "pluginname/*" : "off"
}
```

**Comment by gajus** [comment]
> I assume what they want is a way to cancel out the presence of a plugin in a “plugins” list, as well as setting all rules from that plugin to “off”, such that effectively no code from a plugin is evaluated (obv some would be, since plugin code can create configs, and it might be required somewhere - but the rules wouldn’t be invoked)

That is correct.

**Comment by nzakas** [comment]
Okay, thanks for the clarification. I’m still not seeing a clear use case for disabling a plugin. If a shared config is using a plugin you don’t want, it seems like that’s not the right config to extend from. Configs aren’t meant to be infinitely configurable, and were designed specifically to override or add, not to delete. 

Maybe someone else on the team feels differently, but I’m not in favor of this.

**Comment by nzakas** [comment]
**TSC Summary:** This proposal seeks a way to disable a plugin that has been included by a shared config. Essentially, a way to say "I want the shared config *except* for this plugin it's including." Disabling the plugin would mean disabling all rules, unregistering processors, removing environments, etc.

**TSC Question:** Shall we accept this proposal?

**Comment by btmills** [comment]
Thanks for this suggestion, @gajus. We discussed it in today's TSC meeting and came to the conclusion that the implementation complexities here would be hard to get right. Since there are alternative ways to handle this use case, we'd rather not add this to ESLint's config system. We're working on a new config system that'll hopefully make plugins and everything else easier, and if you're interested, you can follow along in https://github.com/eslint/eslint/issues/13481.
